README.md

jQuery SlideOutPanel
-------

[![NPM_PACKAGE](https://img.shields.io/badge/NPM%20-Package-%23cb3837)](https://www.npmjs.com/package/jquery-slideoutpanel)

Simple jQuery Plugin to add a sliding out panel.


### Installation


```
npm install --save-dev jquery-slideoutpanel
```


### Demo

[https://webdevnerdstuff.github.io/jquery-SlideOutPanel/demo.html](https://webdevnerdstuff.github.io/jquery-SlideOutPanel/demo.html)

<br>

## Usage

### HTML

The html must be set up with the proper id and internal elements.

Allowed tags: `<header>` `<section>` `<footer>`

Optional tags: `<header>` `<footer>`

Required tag: `<section>`

```html
<div id="slide-out-panel" class="slide-out-panel">
    <header>Panel Title</header>
    <section>Panel content section</section>
    <footer>Panel footer</footer>
</div>
```


### Options

<br>

Name                | Type    | Default | Options | Description
:-----              | :------ | :-----  | :-----  | :-----
bodyPush            | boolean | false   | true <br> false| Enables pushing the body when the tab opens (works with slideFrom right and left
breakpoint          | String  | 768px   | | Sets the breakpoint (matches the breakpoint in SCSS/CSS). This is used when bodyPush is true
closeBtn            | html    |  `&#10005;` <br> &#10005;  | |  Add a custom close button instead of the default
closeBtnSize        | String  | 12px    | |  Adjust the close button size
enableEscapeKey     | Boolean | false   | true <br> false |  Enables the esc key to close all panels
offsetTop           | String  | 0       | |  Offset the top of the panel
screenClose         | Boolean | true    | true <br> false |  Enables closing of panels by clicking on the background screen.
screenOpacity       | String  | 0.5     | |  Set the background screen's opacity
screenZindex        | String  | 9998    | |  Set the background screen's z-index
showScreen          | Boolean | true    | true <br> false |  Enable/Disable showing the background screen
slideFrom           | String  | right   | top <br> right <br> bottom <br> left |  Set to choose where the panel should slide out from
transition          | String  | ease    | | Set the transition-timing-function. Accepts the standard values used with CSS.
transitionDuration  | String  | 0.35s   | | Set the duration of the transitions. Adding "s" is optional.
width               | String  | 350px   | |  Set the panels width

<br>

```javascript
$('#slide-out-panel').SlideOutPanel({
    bodyPush: false,
    closeBtn: '<i class="fas fa-times"></i>',
    closeBtnSize: '',
    enableEscapeKey: true,
    offsetTop: '50px',
    screenClose: false,
    screenOpacity: '1',
    screenZindex:  '9998',
    showScreen: false,
    slideFrom: 'right',
    transition: 'ease',
    transitionDuration: '0.35s',
    width: '350px',
});
```


### Events

<br>

Name          | Description
:-----        | :-----
rendered      | Fired after the panel is finished building
beforeOpen    | Fired before panel opens
afterOpen     | Fired after panel has opened
beforeClosed  | Fired before panel is closed
afterClosed   | Fired after the panel is closed

<br>

```javascript
$('#slide-out-panel').SlideOutPanel({
    rendered() {
      // Some code...
    },
    beforeOpen() {
      // Some code...
    },
    afterOpen() {
      // Some code...
    },
    beforeClosed() {
      // Some code...
    },
    afterClosed() {
      // Some code...
    },
});
```


### Methods

<br>

Name    | Description
:-----  | :-----
open    | Opens the panel
close   | Closes the panel
toggle  | Toggles the panel open/close
destroy | Removes the panel from the DOM

<br>

```javascript
const slideOutPanel = $('#slide-out-panel').SlideOutPanel();

slideOutPanel.open();

slideOutPanel.close();

slideOutPanel.toggle();

slideOutPanel.destroy();
```


### Sass Variables

<br>

Variable                      | Default   | Description
:-----                        | :-----    | :-----
$pieces-padding               | 15px      | Padding for the panel pieces (header, section, footer)
$so-screen-sm                 | 768px     | Responsive breakpoint
$so-close-btn-color           | #000      | Color of the close button
$so-close-font-size           | 12px      | Font size of the close button
$so-container-background      | #fff      | The panel background color
$so-container-box-shadow      | -3px 3px 9px rgba(0, 0, 0, .3) | The panel box shadow
$so-container-transition      | top ease, right ease, bottom ease, left ease | Transition effect
$so-container-z-index         | 9999      | The panel z-index
$so-content-no-header-padding-top | $pieces-padding  * 2 | The top padding of the `<section>` when there is no header
$so-header-background-color   | #fff     | The header background color
$so-header-border-color       | #e5e5e5  | The header bottom border color
$so-header-border-width       | 1px      | The header border width
$so-content-background-color  | #fff     | The `<section>` background color
$so-footer-background-color   | #fff     | The `<footer>` background color

<br>

## Examples

Initialize plugin:
```javascript
$('#slide-out-panel').SlideOutPanel();
 ```

Changing the direction the panel slides out from:
```javascript
$('#slide-out-panel').SlideOutPanel({
    slideFrom: 'top',
});

$('#slide-out-panel').SlideOutPanel({
    slideFrom: 'right',
});

$('#slide-out-panel').SlideOutPanel({
    slideFrom: 'bottom',
});

$('#slide-out-panel').SlideOutPanel({
    slideFrom: 'left',
});
```

Close button customization:
```javascript
$('#slide-out-panel').SlideOutPanel({
    closeBtn: '<i class="fas fa-times"></i>',
    closeBtnSize: '18px',
});
```

Background screen customization:
```javascript
$('#slide-out-panel').SlideOutPanel({
    screenClose: true,
    screenOpacity: '0.5',
    screenZindex: '9998',
    showScreen: true,
});
```

<br>

## Dependencies

jQuery

<br>

## Change Log

[CHANGELOG](https://github.com/webdevnerdstuff/jquery-SlideOutPanel/blob/master/CHANGELOG.md)

<br>

## License

Copyright (c) 2020 WebDevNerdStuff
Licensed under the MIT license.

[![@WebDevNerdStuff](https://img.shields.io/badge/github-webdevnerdstuff-brightgreen.svg)](https://github.com/webdevnerdstuff)


CHANGELOG.md
# Change Log
All notable changes to the "jquery-SlideOutPanel" plugin will be documented in this file.

## v1.0.31
Mar 7, 2020
<br>
[master] Fixing iPad issue with close screen not working properly. (webdevnerdstuff)

## v1.0.30
Mar 7, 2020
<br>
[master] Adding touch event for mobile devices to close panel. (webdevnerdstuff)

## v1.0.28
Mar 7, 2020
<br>
[master] Adjust starting position so box-shadow is not on screen. (webdevnerdstuff)

## v1.0.27
Mar 7, 2020
<br>
[master] Fixed issue where multiple elements using bodyPush conflicting with each other. (webdevnerdstuff)

## v1.0.26
Mar 5, 2020
<br>
[master] Fixed esc key closing always being on. (webdevnerdstuff)

## v1.0.25
Mar 5, 2020
<br>
[master] Added box-sizing for elements within the slide-out-panel (webdevnerdstuff)

## v1.0.21
Mar 4, 2020
<br>
[master] Fixed issue with the close click event unbinding all body click events (webdevnerdstuff)

## v1.0.20
Mar 4, 2020
<br>
[master] Fixed issue with bodyPush affecting non bodyPush when dealing with responsiveness (webdevnerdstuff)

## v1.0.18
Mar 4, 2020
<br>
[master] Fixed some issues with the responsiveness when using bodyPush (webdevnerdstuff)

## v1.0.0
Feb 27, 2020
<br>
[master] initial release (webdevnerdstuff)


Gruntfile.js
const sass = require('node-sass');

module.exports = grunt => {
	require('load-grunt-tasks')(grunt);

	// Project configuration.
	grunt.initConfig({
		pkg: grunt.file.readJSON('package.json'),
		name: 'slide-out-panel',
		// ---------------------------------------------------- CLEAN //
		clean: {
			dist: {
				src: ['dist/*'],
			},
			temp: {
				src: ['_temp'],
			},
			demo: ['docs/css', 'docs/js'],
		},
		// ---------------------------------------------------- COPY //
		copy: {
			main: {
				files: [
					{
						expand: true,
						cwd: '_temp/',
						src: '**',
						dest: 'dist/js/',
					},
				],
			},
			scss: {
				files: [
					{
						expand: true,
						cwd: 'src/',
						src: '**/*.scss',
						dest: 'dist/scss/',
					},
				],
			},
		},
		// ==================================================== JAVASCRIPT TASKS //
		// ---------------------------------------------------- BABEL //
		babel: {
			options: {
				comments: false,
				presets: ['@babel/preset-env'],
			},
			dev: {
				options: {
					sourceMap: true,
				},
				files: {
					'_temp/<%= name %>.js': 'src/<%= name %>.js',
				},
			},
			dist: {
				options: {
					sourceMap: false,
				},
				files: {
					'_temp/<%= name %>.js': 'src/<%= name %>.js',
				},
			},
		},
		// ---------------------------------------------------- ESLINT - DONE //
		eslint: {
			options: {
				configFile: '.eslintrc.js',
				maxWarnings: 10,
				fix: grunt.option('fix'),
			},
			target: ['src/**/*.js'],
		},
		// ---------------------------------------------------- UGLIFY //
		uglify: {
			options: {
				beautify: false,
				mangle: true,
				sourceMap: false,
			},
			dist: {
				files: {
					'dist/js/<%= name %>.min.js': ['_temp/<%= name %>.js'],
					'docs/js/<%= name %>.min.js': ['_temp/<%= name %>.js'],
				},
			},
		},
		// ==================================================== STYLE TASKS //
		// ---------------------------------------------------- SASS //
		sass: {
			options: {
				implementation: sass,
				precision: 10,
			},
			dev: {
				options: {
					outputStyle: 'expanded',
					sourceMap: true,
				},
				files: {
					'dist/css/<%= name %>.css': 'src/<%= name %>.scss',
					'docs/css/<%= name %>.css': 'src/<%= name %>.scss',
					'docs/css/docs.css': 'src/docs/docs.scss',
				},
			},
			dist: {
				options: {
					outputStyle: 'compressed',
					sourceMap: false,
				},
				files: {
					'dist/css/<%= name %>.min.css': 'src/<%= name %>.scss',
					'docs/css/<%= name %>.min.css': 'src/<%= name %>.scss',
					'docs/css/docs.min.css': 'src/docs/docs.scss',
				},
			},
		},
		// ---------------------------------------------------- SASS-LINT //
		sasslint: {
			options: {
				configFile: '.sass-lint.yml',
				formatter: 'stylish',
			},
			target: ['src/**/*.scss'],
		},
		// ---------------------------------------------------- POSTCSS //
		postcss: {
			options: {
				processors: [
					require('autoprefixer')({
						config: '.browserslistrc',
					}),
				],
			},
			dev: {
				map: {
					inline: false,
					sourcesContent: true,
					prev: 'dist/css/<%= name %>.css.map',
					annotation: 'dist/css/',
				},
				src: 'dist/css/<%= name %>.css',
			},
			dist: {
				src: 'dist/css/<%= name %>.min.css',
			},
			docs: {
				map: {
					inline: false,
					sourcesContent: true,
					prev: 'docs/css/<%= name %>.css.map',
					annotation: 'docs/css/',
				},
				src: 'docs/css/docs.min.css',
			},
			dist_docs_sop: {
				src: 'docs/css/<%= name %>.min.css',
			},
			dist_docs: {
				src: 'docs/css/docs.min.css',
			},
		},
		// ---------------------------------------------------- WATCH //
		watch: {
			options: {
				spawn: false,
			},
			javascript: {
				files: ['src/**/*.js'],
				tasks: ['eslint', 'clean:temp', 'babel', 'uglify', 'clean:temp'],
			},
			scss: {
				files: ['src/**/*.scss'],
				tasks: ['sass', 'postcss', 'sasslint'],
			},
		},
		paths: {
			cssDist: 'dist/css',
			jsDist: 'dist/js',
			sassDist: 'dist/scss',
		},
	});

	// ---------------------------------------------------- LOAD TASKS //
	grunt.loadNpmTasks('gruntify-eslint');

	// ---------------------------------------------------- MAIN TASKS //
	grunt.registerTask('default', ['watch']);
	grunt.registerTask('build', ['clean', 'babel', 'copy', 'uglify', 'sass', 'postcss', 'clean:temp']);
};


src/slide-out-panel.js
(function($, window) {
	const pluginName = 'SlideOutPanel';

	// Create the defaults once //
	const defaults = {
		// Options //
		bodyPush: false,
		breakpoint: '768px',
		closeBtn: '&#10005;',
		closeBtnSize: '',
		enableEscapeKey: false,
		offsetTop: 0,
		screenClose: true,
		screenOpacity: '0.5',
		screenZindex: '9998',
		showScreen: true,
		slideFrom: 'right',
		transition: 'ease',
		transitionDuration: '0.35s',
		width: '350px',
		// Events //
		afterClosed() { },
		afterOpen() { },
		beforeClosed() { },
		beforeOpen() { },
		rendered() { },
	};

	let globalSettings;
	let pluginLoaded = false;

	// ---------------------------------------------------- PLUGIN DEFINITION //
	$.fn[pluginName] = function(options) {
		let componentElm;
		let headerElm;
		let contentElm;
		let footerElm;
		let screenSize;

		// Check to see if element exists //
		if (typeof $(this).attr('id') === 'undefined') {
			console.error(`${pluginName}: Element not found. Element options:`, options);
			return false;
		}

		// ---------------------------------------------------- SlideOutPanel constructor //
		function SlideOutPanel(element, soOptions) {
			// Collect component elements //
			componentElm = $(`#${element.id}`);
			headerElm = $(`#${element.id} header`);
			contentElm = $(`#${element.id} section`);
			footerElm = $(`#${element.id} footer`);

			// Add class if missing to try to prevent showing content before built //
			if (!componentElm.hasClass('slide-out-panel')) {
				componentElm.addClass('slide-out-panel');
			}
			// Check to make sure the content section exists //
			if (typeof $(contentElm).html() === 'undefined') {
				destroyPanel(element.id, `#${element.id} <section> missing.`);
				return false;
			}

			// Set settings //
			globalSettings = globalSettings || {};
			globalSettings[element.id] = element;

			// Merge defaults and options //
			globalSettings[element.id] = $.extend({}, defaults, soOptions);

			// Set transitionDuration to INT //
			let transitionDuration = globalSettings[element.id].transitionDuration;
			transitionDuration = transitionDuration.replace('s', '');
			globalSettings[element.id].transitionDuration = parseFloat(transitionDuration);

			// Window size //
			orientationChange();

			this._defaults = defaults;
			this._name = pluginName;

			// Start building the panel //
			init(element.id, globalSettings[element.id]);
			return false;
		}

		// ---------------------------------------------------- INIT //
		const init = (elmId, settings) => {
			component(elmId, settings);

			if (typeof $(headerElm).html() !== 'undefined') {
				header(elmId, settings);
			}

			content(elmId, settings);
			footer();

			// Check if background screen should be shown //
			if (settings.showScreen) {
				screen(elmId, settings);
			}

			// Call rendered method when finished with setup //
			settings.rendered();
			return false;
		};

		// ---------------------------------------------------- COMPONENT //
		const component = (elmId, settings) => {
			const openHorizontal = settings.slideFrom === 'left' || settings.slideFrom === 'right';

			componentElm
				.addClass(`slide-out-panel-container ${settings.slideFrom}`)
				.css({
					display: 'block',
					height: settings.offsetTop === 0 ? 'initial' : `calc(100vh - ${settings.offsetTop})`,
					top: settings.slideFrom === 'bottom' ? 'initial' : settings.offsetTop,
					transitionTimingFunction: `${settings.transition}`,
					transitionDuration: `${settings.transitionDuration}s`,
					width: settings.width,
				})
				.attr('data-id', elmId);

			if (settings.bodyPush && openHorizontal) {
				$('html').css({
					transition: `width ${settings.transitionDuration}s ${settings.transition}`,
				});
			}

			// Start with the panel hidden //
			// If sliding out from left or right //
			if (openHorizontal) {
				const measure = settings.width.match(/[a-zA-Z]+/);
				let num = settings.width.match(/(\d+)/);
				num = parseInt(num) + 10;

				componentElm.addClass(`slide-out-panel-container ${settings.slideFrom}`).css({
					left: settings.slideFrom !== 'left' ? 'initial' : `-${num}${measure}`,
					right: settings.slideFrom !== 'right' ? 'initial' : `-${num}${measure}`,
				});
			}
			else {
				// If sliding out from top or bottom //
				componentElm.addClass(`slide-out-panel-container ${settings.slideFrom}`).css({
					bottom: settings.slideFrom !== 'bottom' ? 'initial' : '-100%',
					top: settings.slideFrom !== 'top' ? 'initial' : '-100%',
					width: '100vw',
				});
			}
		};

		// ---------------------------------------------------- PANEL PIECES //
		// -------------------------- Header //
		const header = (elmId, settings) => {
			if (typeof headerElm.html() !== 'undefined') {
				headerElm.addClass('slide-out-header').append(closeBtn(elmId, settings));
			}
		};

		// -------------------------- Content //
		const content = (elmId, settings) => {
			contentElm.addClass('slide-out-content');

			if (typeof headerElm.html() === 'undefined') {
				contentElm.append(closeBtn(elmId, settings));
			}
		};

		// -------------------------- Footer //
		const footer = () => {
			footerElm.addClass('slide-out-footer');
		};

		// -------------------------- Close button //
		const closeBtn = (elmId, settings) => {
			let btnHtml = '';
			btnHtml += `<span class="close-slide-out-panel" data-id="${elmId}" ${settings.closeBtnSize ?
				`style="font-size: ${settings.closeBtnSize}; display: block;"` :
				''}>`;
			btnHtml += settings.closeBtn;
			btnHtml += '</span>';

			return btnHtml;
		};

		// ---------------------------------------------------- SCREEN //
		const screen = (elmId, settings) => {
			let screenHtml = '';
			screenHtml += '<div ';
			screenHtml += `id="slide-out-panel-screen-${elmId}"`;
			screenHtml += 'class="slide-out-panel-screen" ';
			screenHtml += `data-id="${elmId}" `;
			screenHtml += 'style="';
			screenHtml += `background-color: rgba(0, 0, 0, ${settings.screenOpacity});`;
			screenHtml += `transition-timing-function: ${settings.transition};`;
			screenHtml += `transition-duration: ${settings.transitionDuration}s;`;
			screenHtml += `z-index: -${settings.screenZindex};`;
			screenHtml += '">';
			screenHtml += '</div>';

			$('body').append(screenHtml);

			// Touch detection for mobile devices //
			if (settings.screenClose) {
				const touchScreen = document.getElementById(`slide-out-panel-screen-${elmId}`);

				touchScreen.addEventListener('touchend', e => {
					e.preventDefault();
					const closeBtnElemId = $(this).attr('data-id');
					closePanel(closeBtnElemId, globalSettings[closeBtnElemId], this);
					return false;
				}, false);
			}
		};

		// ---------------------------------------------------- PANEL POSITIONS //
		const panelPositions = (elmId, settings, action) => {
			const openHorizontal = settings.slideFrom === 'left' || settings.slideFrom === 'right';
			let positionVal;
			let css;

			// Set position value //
			if (action === 'open') {
				positionVal = '0';
			}
			else {
				const measure = settings.width.match(/[a-zA-Z]+/);
				let num = settings.width.match(/(\d+)/);
				num = parseInt(num) + 10;
				positionVal = openHorizontal ? `-${num}${measure}` : '-100%';
			}

			// Slide out from left or right //
			if (openHorizontal) {
				css = {
					left: settings.slideFrom === 'right' ? 'initial' : positionVal,
					right: settings.slideFrom === 'left' ? 'initial' : positionVal,
				};
			}
			else {
				// Slide out from top or bottom //
				css = {
					top: settings.slideFrom === 'bottom' ? 'initial' : positionVal,
					bottom: settings.slideFrom === 'top' ? 'initial' : positionVal,
				};
			}

			if (action === 'open') {
				$(`#${elmId}`).addClass('open');
			}
			else {
				$(`#${elmId}`).removeClass('open');
			}

			$(`#${elmId}`).css(css);
			return false;
		};

		// ---------------------------------------------------- CLOSE PANEL //
		const closePanel = (elmId = false, settings = false, elm = false) => {
			// Check if panel exists //
			if (typeof settings === 'undefined') {
				destroyPanel(elmId, `#${elmId} doesn't exist.`);
				return false;
			}
			else if ($(elm).hasClass('slide-out-panel-screen') && !settings.screenClose) {
				// If screenClose is true, do not close panel when clicked //
				return false;
			}

			// ESC Key Recursion as there is no elmId associated with it //
			if (!elmId && !settings && !elm) {
				// Close all panels //
				Object.entries(globalSettings).forEach(([key, val]) => {
					if (globalSettings[key].enableEscapeKey) {
						closePanel(key, globalSettings[key]);
					}
				});

				return false;
			}

			// Call beforeClosed method before panel is closed //
			settings.beforeClosed();

			// Close Panel and adjust left/right styles //
			panelPositions(elmId, settings, 'close');

			// Hide Screen //
			$(`#slide-out-panel-screen-${elmId}`).removeClass('open').css({
				opacity: 0,
				transitionDuration: `${settings.transitionDuration}s`,
			});

			if (settings.bodyPush && (settings.slideFrom === 'left' || settings.slideFrom === 'right')) {
				$('html').css({
					width: '100%',
				});
			}

			setTimeout(() => {
				// Move screen z-index back //
				$(`#slide-out-panel-screen-${elmId}`).css({
					zIndex: '-9999',
				});

				if (settings.bodyPush && (settings.slideFrom === 'left' || settings.slideFrom === 'right')) {
					$('html').removeClass(`slide-out-${settings.slideFrom}`).css('position', 'initial');
				}

				// Call afterClosed method after panel is closed //
				settings.afterClosed();
			}, settings.transitionDuration * 1000);

			return false;
		};

		// ---------------------------------------------------- DESTROY PANEL //
		const destroyPanel = (elmId, msg = false) => {
			$(`#${elmId}, .slide-out-panel-screen[data-id="${elmId}"]`).remove();

			// Check for message to log error //
			if (msg) {
				console.error(`${pluginName}: ${msg}`);
			}

			return false;
		};

		// ---------------------------------------------------- OPEN PANEL //
		const openPanel = (elmId = false, settings) => {
			// Check if panel exists //
			if (typeof settings === 'undefined') {
				destroyPanel(elmId, `#${elmId} doesn't exist.`);
				return false;
			}

			if (settings.bodyPush && (settings.slideFrom === 'left' || settings.slideFrom === 'right')) {
				$('html').addClass(`slide-out-${settings.slideFrom}`);
			}

			// Call beforeOpen method before panel is open //
			settings.beforeOpen();

			// Open Panel and adjust left/right styles //
			panelPositions(elmId, settings, 'open');

			// Show Screen //
			$(`#slide-out-panel-screen-${elmId}`).addClass('open').css({
				opacity: 1,
				transitionDuration: `${settings.transitionDuration}s`,
				zIndex: `${settings.screenZindex}`,
			});

			if (settings.bodyPush && (settings.slideFrom === 'left' || settings.slideFrom === 'right')) {
				$('html').css({
					position: 'absolute',
					width: `calc(${$('body').width()}px - ${settings.width}`,
				});
			}

			// Call afterOpen method after panel is opened //
			setTimeout(() => {
				settings.afterOpen();
			}, settings.transitionDuration * 1000);

			return false;
		};

		// ---------------------------------------------------- TOGGLE PANEL //
		const togglePanel = (elmId = false) => {
			if ($(`#${elmId}`).hasClass('open')) {
				closePanel(elmId, globalSettings[elmId]);
			}
			else {
				openPanel(elmId, globalSettings[elmId]);
			}

			return false;
		};

		// ---------------------------------------------------- WINDOW SIZE & ORIENTATION CHANGING //
		const orientationChange = () => {
			const screenWidth = $(window).width();

			for (let i = 0; i < Object.values(globalSettings).length; i += 1) {
				const settings = Object.values(globalSettings)[i];
				const breakpoint = settings.breakpoint.match(/(\d+)/);

				if (screenWidth < breakpoint[0] && settings.bodyPush) {
					$('html').addClass('slide-out-panel-static');
					break;
				}
				else {
					$('html').removeClass('slide-out-panel-static');

					if (settings.bodyPush && $(`.slide-out-panel-container.${settings.slideFrom}`).hasClass('open')) {
						$('html').css({
							width: `calc(${screenWidth}px - ${settings.width}`,
						});
					}
				}
			}
		};

		// Resize //
		let resizeTimer;

		$(window).on('resize', e => {
			clearTimeout(resizeTimer);
			resizeTimer = setTimeout(() => {
				orientationChange();
			}, 250);
		});

		// ---------------------------------------------------- Public Functions //
		// -------------------------- Close Panel //
		this.close = function() {
			const closeElemId = $(this).attr('id');
			closePanel(closeElemId, globalSettings[closeElemId]);
		};

		// Click to close //
		if (!pluginLoaded) {
			$('body').on('click', '.close-slide-out-panel, .slide-out-panel-screen', function() {
				const closeBtnElemId = $(this).attr('data-id');
				closePanel(closeBtnElemId, globalSettings[closeBtnElemId], this);
				return false;
			});

			pluginLoaded = true;
		}

		// -------------------------- Escape Key to Close
		$('body').on('keyup', e => {
			if (e.keyCode === 27) {
				closePanel();
			}
		});

		// -------------------------- Destroy Panel //
		this.destroy = () => {
			const destroyElemId = $(this).attr('id');
			destroyPanel(destroyElemId);
		};

		// -------------------------- Open Panel //
		this.open = function() {
			const openElemId = $(this).attr('id');
			if (!$(`#${openElemId}`).hasClass('open')) {
				openPanel(openElemId, globalSettings[openElemId]);
			}
		};

		// -------------------------- Toggle Panel //
		this.toggle = function() {
			const openElemId = $(this).attr('id');
			togglePanel(openElemId, globalSettings[openElemId]);
		};

		// ---------------------------------------------------- Set Plugin //
		return this.each(function() {
			$.data(this, `plugin-${pluginName}-${$(this).attr('id')}`, new SlideOutPanel(this, options));
		});
	};
}(jQuery, window));
